### PR TITLE
Update base apptainer, clang, gcc, libmesh, petsc, wasp

### PR DESCRIPTION
**Pull request by loganharbour**

### CODE REVIEW

**Comment by cticenhour** [comment]
image

File: scripts/update_and_rebuild_libmesh.sh [hunk 1]
```diff
@@ -151,16 +158,16 @@ if [ -z "$go_fast" ]; then
...
   mkdir -p $LIBMESH_BUILD_DIR
   cd $LIBMESH_BUILD_DIR
 
   source $SCRIPT_DIR/configure_libmesh.sh
-  SRC_DIR=${SCRIPT_DIR}/../libmesh configure_libmesh $DISABLE_TIMESTAMPS \
-                                                     $VTK_OPTIONS \
-                                                     $* | tee -a "$SCRIPT_DIR/$DIAGNOSTIC_LOG" || exit 1
+  SRC_DIR=${LIBMESH_SRC_DIR} configure_libmesh $DISABLE_TIMESTAMPS \
+                                               $VTK_OPTIONS \
+                                                $* | tee -a "$SCRIPT_DIR/$DIAGNOSTIC_LOG" || exit 1
```
Comment (cticenhour) [inline scripts/update_and_rebuild_libmesh.sh:170]:
```suggestion
  SRC_DIR=${LIBMESH_SRC_DIR} configure_libmesh $DISABLE_TIMESTAMPS \
                                               $VTK_OPTIONS \
                                               $* | tee -a "$SCRIPT_DIR/$DIAGNOSTIC_LOG" || exit 1
```

File: scripts/update_and_rebuild_petsc.sh [hunk 2]
```diff
@@ -74,21 +80,22 @@ if [[ -n "$go_fast" && $# != 1 ]]; then
...
-    echo "git submodule command failed, are your proxy settings correct?"
-    exit 1
+if [ -z "$skip_sub_update" ]; then
+  cd ${MOOSE_DIR}
+  git_dir=`git rev-parse --show-cdup 2>/dev/null`
+  if [[ -z "$go_fast" && $? == 0 && "x$git_dir" == "x" ]]; then
+    git submodule update --init --recursive petsc
+    if [[ $? != 0 ]]; then
+      echo "git submodule command failed, are your proxy settings correct?"
+      exit 1apptainer shell
```
Comment (cticenhour) [inline scripts/update_and_rebuild_petsc.sh:97]:
```suggestion
      exit 1
```

Typo?

File: conda/moose-dev/conda_build_config.yaml [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
-  - moose-libmesh 2023.11.06 build_1
+  - moose-libmesh 2023.11.28 build_0
 
 moose_wasp:
-  - moose-wasp 2023.11.06
+  - moose-wasp 2023.11.28
 
 moose_tools:
-  - moose-tools 2023.10.27
+  - moose-tools 2023.11.28
```
Comment (cticenhour) [inline conda/moose-dev/conda_build_config.yaml:8]:
```suggestion
  - moose-tools 2023.10.27
```

`moose-tools` is not being updated in this PR.

**Review by cticenhour — CHANGES_REQUESTED** [review]
Reviewing only conda, configure scripts, and related changes. Leaving apptainer to @milljm.

**Review by lindsayad — COMMENTED** [review]
wow we went from no PETSc updates in years to three updates in a few months! I'm not complaining!

File: apptainer/moose-dev.def [hunk 4]
```diff
@@ -71,23 +77,16 @@ From: {{ APPTAINER_FROM }}
     # Create list of Conda Packages to install
     MAMBA_PACKAGES="moose-tools=${MOOSE_TOOLS_VERSION}"
     # Install packages
-    mamba create -yq -n ${MOOSE_MAMBA_ENV} python=${PYTHON_VERSION} ${MAMBA_PACKAGES}
+    mamba create -yq -n ${MOOSE_CONDA_ENV_NAME} python=${PYTHON_VERSION} ${MAMBA_PACKAGES} openssl=3.1.4
```
Comment (milljm) [inline apptainer/moose-dev.def:80]: pesky openssl.

I am a little confused why I need to pin this while we build Conda packages for Mac targets. But we need this here for runtime Linux targets =D

Comment (loganharbour) [reply apptainer/moose-dev.def:80]: I don't think we need this anymore... but we're in too deep and can remove it on the next update

File: apptainer/mpich.def [hunk 5]
```diff
@@ -159,9 +162,9 @@ EOF
     cmake ../llvm -G 'Unix Makefiles' \
       -DCMAKE_BUILD_TYPE=Release \
       -DLLVM_ENABLE_PROJECTS='clang;clang-tools-extra;compiler-rt;libcxx;libcxxabi;libunwind;openmp;lldb' \
-      -DCMAKE_INSTALL_PREFIX=/opt/llvm-${MIN_CLANG_VERSION}  \
-      -DCMAKE_INSTALL_RPATH:STRING=/opt/llvm-${MIN_CLANG_VERSION}/lib \
-      -DCMAKE_INSTALL_NAME_DIR:STRING=/opt/llvm-${MIN_CLANG_VERSION}/lib \
+      -DCMAKE_INSTALL_PREFIX=/opt/llvm  \
+      -DCMAKE_INSTALL_RPATH:STRING=/opt/llvm/lib \
+      -DCMAKE_INSTALL_NAME_DIR:STRING=/opt/llvm/lib \
```
Comment (milljm) [inline apptainer/mpich.def:167]: why didn't I think of that...

File: scripts/apptainer_generator.py [hunk 6]
```diff
@@ -448,35 +495,56 @@ def create_filename(app_root, section_key, actions):
...
+                continue
+
+            var = f'SECTION_{section.upper()}'
+            file_contents = open(full_filename, 'r').read()
+            contents = f"# Begin include from '{filename}' in {app_name}\n"
+            contents += file_contents
+            contents += f"# End include from '{filename}' in {app_name}\n"
+            contents = self.add_def_whitespace(contents)
+
+            jinja_data[var] = contents
```
Comment (milljm) [inline scripts/apptainer_generator.py:547]: So if I am reading this right, Application owners can introduce their own scripts? Neat!

Comment (loganharbour) [reply scripts/apptainer_generator.py:547]: yeah, I finished that up in conjunction with getting the cardinal container working: https://github.com/loganharbour/cardinal/tree/813_add_context/apptainer

**Review by milljm — APPROVED** [review]
heck yeah! 👍

File: apptainer/moose-dev.def (hunk 4 above)
Comment (loganharbour) [reply apptainer/moose-dev.def]: Went ahead and removed this

**Review by cticenhour — APPROVED** [review]
Conda changes look good.

**Review by milljm — APPROVED** [review]
noice

**Comment by loganharbour** [comment]
Failed and cancelled tests will no longer be used after this merge.

**Pull request merged**